### DISCUSSION TITLE
Create multilabel dataset

### DISCUSSION THREAD

**Discussion by kirrog** [Project-MONAI/MONAILabel#1225] [General]
Hi everyone!

We are trying to learn deepedit, using monai radilogy app as template, to segment many labels, about 20.
We've changed 
self.labels in configs/deepedit to vocabulary contains pairs of label_name:label_number, similary to original
And place in root of "studies" directory all CBCT as nifti
In root/labels/final we've placed nifti in which we place segmentation all labels, as label_num in voxels where segmentation should be
We try to learn on all segments and found out, what if one label is huge, and other is tiny, deepedit prefer to segment the huge one. It's logical because of loss. So we tried to segment tiny ones by one, but monai used all segmentation map as segmentation of this label.
We tried to segment many labels, but not all, and when got results near what did we expect, but now we fear of probability, what monai may do some missunderstanding of our data, and so we have worser results, when we may, because of this missunderstangind

Can you please explain:
1) should we create new nifti for each set of labels, or just for one-label we should create one-label map?
2) If we train for multilabel, we can delete some labels from deepedit config and monai would use only written in config labels for learning, and ignore the labels that exists in label-map nifti?

**Comment by diazandr3s** [comment]
> Hi everyone!
> 
> We are trying to learn deepedit, using monai radilogy app as template, to segment many labels, about 20. We've changed self.labels in configs/deepedit to vocabulary contains pairs of label_name:label_number, similary to original And place in root of "studies" directory all CBCT as nifti In root/labels/final we've placed nifti in which we place segmentation all labels, as label_num in voxels where segmentation should be We try to learn on all segments and found out, what if one label is huge, and other is tiny, deepedit prefer to segment the huge one. It's logical because of loss. So we tried to segment tiny ones by one, but monai used all segmentation map as segmentation of this label. We tried to segment many labels, but not all, and when got results near what did we expect, but now we fear of probability, what monai may do some missunderstanding of our data, and so we have worser results, when we may, because of this missunderstangind
> 
> Can you please explain:
> 
>     1. should we create new nifti for each set of labels, or just for one-label we should create one-label map?
> 
>     2. If we train for multilabel, we can delete some labels from deepedit config and monai would use only written in config labels for learning, and ignore the labels that exists in label-map nifti?

Hi @kirrog,

It is great to hear you're using MONAI Label for this type of images. Is it for teeth and mandible segmentation?
One of the great things about DeepEdit is that it allows interactive segmentation. However, one of the disadvantages of this model is that it uses the whole image for training and inference. This means there is a downsampling process during training and inference to fit into the GPU.

I'm working with @dgmato on the same task. would you like to join forces? We have around 40 volumes fully segmented for this.

For this task we're using the multistage approach developed for vertebra segmentation: https://github.com/Project-MONAI/MONAILabel/tree/main/sample-apps/radiology#multistage-vertebra-segmentation

With regards to your questions:

> 1. should we create new nifti for each set of labels, or just for one-label we should create one-label map?

For a multilabel segmentation task, MONAI Label needs one file containing one label map with all classes/labels.

> 2. If we train for multilabel, we can delete some labels from deepedit config and monai would use only written in config labels for learning, and ignore the labels that exists in label-map nifti?

Yes. MONAI Label only considers the labels defined in the label dictionary. A similar use case is here: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/configs/deepedit.py#L42-L51
where we used only some of the labels in the BTCV dataset.

Hope this helps,

**Reply by dalvSilvermann** [reply]
Hi @diazandr3s
Can you answer on questions below, please:

1. Does the label map with all classes/labels must be the same count/size of Scalar Range steps quantity for all files containing in dataset? For example, if we working with first five labels, does it matter, that in one file the maximum of Scalar Range is equal to 20, and in another is equal to 15, if we are just working with the first five? (0 is background)

2. Does it matter that Scalar Type is float? Not short.

3. Background stays in the last position in dict or any position? =)

Thanks a lot for support!
<img width="1280" alt="Screenshot 2022-12-21 at 19 13 21" src="https://user-images.githubusercontent.com/108551361/209402341-7f56805d-f813-4007-8884-00cea627e69b.png">

**Reply by diazandr3s** [reply]
> Hi @diazandr3s Can you answer on questions below, please:
> 
>     1. Does the label map with all classes/labels must be the same count/size of Scalar Range steps quantity for all files containing in dataset? For example, if we working with first five labels, does it matter, that in one file the maximum of Scalar Range is equal to 20, and in another is equal to 15, if we are just working with the first five? (0 is background)
> 
>     2. Does it matter that Scalar Type is float? Not short.
> 
>     3. Background stays in the last position in dict or any position? =)
> 
> 
> Thanks a lot for support! <img alt="Screenshot 2022-12-21 at 19 13 21" width="1280" src="https://user-images.githubusercontent.com/108551361/209402341-7f56805d-f813-4007-8884-00cea627e69b.png">

Hi @dalvSilvermann,

> 1. Does the label map with all classes/labels must be the same count/size of Scalar Range steps quantity for all files containing in dataset? For example, if we working with first five labels, does it matter, that in one file the maximum of Scalar Range is equal to 20, and in another is equal to 15, if we are just working with the first five? (0 is background)

No, the label maps can have different scalar range. I understand not all the patients have the same number of teeth. As long as the segments have the same index it shouldn't affect the training. i.e. teeth 1 should always have the same index/scalar.

> 2. Does it matter that Scalar Type is float? Not short.

No, it doesn't matter

> 3. Background stays in the last position in dict or any position?

By convention, I put the background as the last segment. But it shouldn't affect if placed in a different position.

Hope this helps,

**Reply by dalvSilvermann** [reply]
Thanks a lot for your reply!
